File: Algorithms/w3_sort/QuickSort/Quicksort.cc
#include<iostream>
#include<vector>
#include<algorithm>
#include<string>
using namespace std;
class Quicksort
{
public:
    int partition(vector<int>& nums, int lo, int hi)
    {
        int i = lo;
        int j = hi + 1;
        while(1)
        {      
            while(nums[lo] > nums[++i])
                if(i == hi)     break;
            while(nums[lo] < nums[--j])
                if(j == lo)     break;
            
            if( i >= j )    break;
            swap(nums[i], nums[j]);

        }
        swap(nums[lo], nums[j]);
        return j;
        
    }
    void sort(vector<int>& nums, int lo, int hi)
    {
        if(lo >= hi)    return;
        int mid = partition(nums, lo, hi);
        sort(nums, lo, mid - 1);
        sort(nums, mid + 1, hi);
    }

    
};

int main()
{
    vector<int> nums;
    Quicksort q;
    while(1)
    {
        string tmp;
        cout<<"Input number: "<<endl;
        cin>>tmp;
        if(!tmp.compare("quit"))   break;
        nums.push_back(stoi(tmp));
    }
    cout<<"---------------------------------"<<endl;
    q.sort(nums, 0, nums.size() - 1);
    for(auto itr = nums.begin(); itr != nums.end(); itr++)
    {
        cout<< *itr <<endl;
    }
}

File: Algorithms/w3_sort/QuickSort/quicksort.java
import java.util.Scanner;

public class Quicksort 
{
    public int partition(int[] nums, int lo, int hi)
    {
        int i = lo;
        int j = hi + 1;
        while(true)
        {
            while(nums[++i] < nums[lo])   
                if(i == hi) break;

            while(nums[--j] > nums[lo])
                if(j == lo) break;
                
            if(i >= j)  break;
            swap(nums, i, j);
        }
        swap(nums, lo, j);
        return j;

    }
    public void swap(int[] nums, int i, int j)
    {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public void sort(int[] nums, int lo, int hi)
    {
        if(lo >= hi)    return;
        int mid = partition(nums, lo, hi);
        sort(nums, lo, mid - 1);
        sort(nums, mid + 1, hi);
    }

    public static void main(String[] args)
    {
        Scanner input = new Scanner(System.in);
        Quicksort q = new Quicksort();
        System.out.println("How many numbers do you wanna sort?\n");
        int m = input.nextInt();
        int[] nums = new int[m];
        int i = 0;
        for(int j = 0; j < m; j++)
        {
            System.out.println("Insert numbers you wanna sort:\n");
            int p = input.nextInt();
            nums[j] = p;

        }
        input.close();
        q.sort(nums, 0, nums.length - 1);
        for(int x : nums)
        {
            System.out.println(x + " ");
        }
    }
    
}
